src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs

```diff
@@ -58,22 +61,76 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
 
         public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder) where T : IWebJobsStartup, new()
         {
-            return builder.UseWebJobsStartup(typeof(T));
+            return builder.UseWebJobsStartup<T>(NullLoggerFactory.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder, ILoggerFactory loggerFactory) where T : IWebJobsStartup, new()
+        {
+            return builder.UseWebJobsStartup(typeof(T), loggerFactory);
         }
 
         public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType)
+        {
+            return builder.UseWebJobsStartup(startupType, NullLoggerFactory.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType, ILoggerFactory loggerFactory)
         {
             if (!typeof(IWebJobsStartup).IsAssignableFrom(startupType))
             {
                 throw new ArgumentException($"The {nameof(startupType)} argument must be an implementation of {typeof(IWebJobsStartup).FullName}");
             }
 
             IWebJobsStartup startup = (IWebJobsStartup)Activator.CreateInstance(startupType);
-            startup.Configure(builder);
 
+            if (loggerFactory == NullLoggerFactory.Instance)
+            {
+                startup.Configure(builder);
+            }
+            else
+            {
+                ConfigureAndLogUserConfiguredServices(startup, builder, loggerFactory);
+            }
             return builder;
         }
 
+        private static void ConfigureAndLogUserConfiguredServices(IWebJobsStartup startup, IWebJobsBuilder builder, ILoggerFactory loggerFactory)
+        {
+            var logger = loggerFactory.CreateLogger<TrackedServiceCollection>();
+
+            if (builder.Services is ITrackedServiceCollection tracker)
+            {
+                if (tracker != null)
+                {
+                    startup.Configure(builder);
+                    StringBuilder sb = new StringBuilder("Services registered by external startup type " + startup.GetType().ToString() + ":");
+
+                    foreach (ServiceDescriptor service in tracker.TrackedCollectionChanges)
+                    {
+                        sb.Append($"\n {service.ServiceType.FullName}: ");
```

Comment:
Does this appear properly formatted? Should it be "\r\n"? https://docs.microsoft.com/en-us/dotnet/api/system.environment.newline?view=netframework-4.8#property-value.

Reply:
Yes, you are right! I will create a bug fix.